### ISSUE TITLE
cannot stub request anymore

### ISSUE THREAD

**Issue by Nghi93** [ashkan18/graphlient#100]
**Describe the bug**
Since 0.7.0 it is not possible to stub_request in ones specs anymore:
<NoMethodError: undefined method `values_at' for #<String:0x00007f9defb82e20>

**To Reproduce**
```
let(:response) do
      {
        'data' => {
          'someQuery' => {
            '__typename' => 'SomeType',
            'id' => 'A123'
          }
        }
      }.to_json
end

before do
  client = Graphlient::Client.new(url, schema_path: x)
  stub_request(:post, url).to_return(status: 200, body: response)
end

it do
  client.execute(some_query)
end
```
We get <NoMethodError: undefined method `values_at' for #<String:0x00007f9defb82e20>

I could track it down to graphql-client-0.18.0/lib/graphql/client.rb:358:
```
358         execute.execute(
359           document: document,
360           operation_name: operation.name,
361           variables: variables,
362           context: context
363        )
```

On a regular call this will result a hash. During a stubbed call it will return my mock "response" string, which does not have the method `values_at` which causes the error above.

I am not sure if this is a bug in graphlient or graphql-client. Could you provide further informations, please?

**Comment by ashkan18** [comment]
Hi, thanks for reporting this! are you still dealing with this issue?

**Comment by Nghi93** [comment]
Hi @ashkan18,

yes, it's still an issue.

**Comment by ashkan18** [comment]
Sorry to hear that, i haven't had a chance to look into it! but feel free to look into it and open a PR if you had a chance.

**Comment by mikecx** [comment]
In case anyone else runs into this, I was able to get around it by adding a content-type header to the stub-request call like so:

```
stub_request(:post, MyGem::Client::ENDPOINT).to_return(status: 200, body: json_response.to_json, headers: { 'Content-Type' => 'application/json' })
```

**Comment by dblock** [comment]
@mikecx Do we need to document this? Care to PR?

**Comment by mikecx** [comment]
@dblock Yeah, seems like it would be good to update the README.md with the updated stubbing. I'll put together a doc update PR when I get some free time, though it might be a few days.

**Comment by ashkan18** [comment]
thanks @mikecx ❤️ 💜 💚

**Comment by taylorthurlow** [comment]
I just upgraded from 0.5 to 0.7 and ran into the same problem. It's possible that I haven't thought this through fully - I think it's related to changes that happened surrounding the Faraday upgrade and how the JSON middleware is used. The bottom line is that it seems like Graphlient should not be assuming that a response body from a graphql API is JSON. A normally operating one would be, but there is no graceful failure here. Especially in the case where the content type of the response is not `application/json`.

I don't quite have the time to dig further right now, though, sorry.

**Comment by taylorthurlow** [comment]
Ok, looks like I had the time 😆 - I think this is the result of a change in the Faraday json response middleware from 1.x to 2.x.

Here is the `faraday_middleware` implementation of the response JSON parser (i.e. how Faraday 1.x works):

https://github.com/lostisland/faraday_middleware/blob/e68ff84470705c8a59d9aa99b98cb36101fa10ad/lib/faraday_middleware/response/parse_json.rb#L13

Regardless of content type, if the body contains any non-whitespace characters, it'll parse it as JSON. Presumably this gets caught somewhere and raised as a `Faraday::ParsingError`.

Contrast this with the Faraday 2.x built-in JSON response middleware:

https://github.com/lostisland/faraday/blob/074506e67c20e5c79f7f8fb544a318eb932d6bc8/lib/faraday/response/json.rb#L24

We `#process_response` (by default call `JSON.parse`) if `#parse_response?` is `true`. It would be false if the content type did not match `/\bjson$/`.

---

So to summarize this is just an (undocumented?) change in how the faraday JSON response middleware works and a change should probably be made to Graphlient to raise an appropriate exception if the response from the GraphQL API is not valid JSON.

**Comment by ashkan18** [comment]
Thanks @taylorthurlow for investigative work 🔬 work! do you mind opening a PR with that proposed change when have a chance?

**Comment by taylorthurlow** [comment]
@ashkan18 Opened 👍
